src/firmware/app/control/python_controller/difference_equation.py

```diff
@@ -0,0 +1,109 @@
+        """
+        self.__numerator_coefficients = tf_numerator_coefficients
+        self.__denominator_coefficients = tf_denominator_coefficients
+
+        # Store the order of the numerator and denominator for calculation later
+        self.__controller_rank = len(self.__denominator_coefficients)
+        self.__input_coeffs_order = len(self.__numerator_coefficients)
+
+        # Normalize by the coefficient of the highest order denominator term
+        # This is done so that the output of the difference equation is not scaled
+        for i in range(0, len(self.__numerator_coefficients)):
+            self.__numerator_coefficients[i] /= self.__denominator_coefficients[0]
+
+        for i in range(0, len(self.__denominator_coefficients)):
+            self.__denominator_coefficients[i] /= self.__denominator_coefficients[0]
+
+        # We must keep track of the system state and previous states
+        self.time_step_count = 0
+        self.__previous_input = []
+        self.__previous_output = []
+
+    def tick(self, current_input):
+
+        """ Ticks the difference equation to the next time step with the current input
+
+        :float current_input: The current input to the discrete system
+        """
+
+        # Generate the systems response due to previous inputs
+        effect_of_previous_input = 0
+
+        # Loop through each coefficient of the numerator
+        for i in range(0, self.__input_coeffs_order):
+
+            # Calculate how many input 'steps' back in time we must use for the given numerator coefficient
+            index = self.time_step_count + (
+                self.__input_coeffs_order - self.__controller_rank - i
+            )
+
+            # If we must go more time steps backwards than are recorded, the term has no effect on the output
+            if index < 0 or len(self.__previous_input) < index + 1:
+                effect_of_previous_input += 0
+            else:
+                effect_of_previous_input += (
+                    self.__numerator_coefficients[i] * self.__previous_input[index]
+                )
+
+        # Generate the systems response due to previous output
+        effect_of_previous_output = 0
+
+        # Loop through each coefficient of the denominator other than the highest order term
+        # as it represents the output of the tick
+        for i in range(1, self.__controller_rank):
+
+            # Calculate the number of time steps backwards to use
+            index = self.time_step_count - i
+
+            # If we must go more time steps backwards than are recorded, the term has no effect on the output
+            if index < 0 or len(self.__previous_output) < index + 1:
+                effect_of_previous_output += 0
+            else:
+                effect_of_previous_output += (
+                    self.__denominator_coefficients[i] * self.__previous_output[index]
+                )
```

Comment:
Consider using filter/map/reduce and list comprehension on lists. Certain for loops are considered [unpythonic](https://docs.quantifiedcode.com/python-anti-patterns/readability/using_an_unpythonic_loop.html)

Reply:
See above comment 👍